### PR TITLE
EVP: don't call memset for evp_pkey_downgrade

### PR DESCRIPTION
**Pull request by danbev**
This commit tries to address a locking issue in `evp_pkey_reset_unlocked`
which can occur when it is called from `evp_pkey_downgrade`.

`evp_pkey_downgrade` will acquire a lock for `pk->lock` and if successful
then call `evp_pkey_reset_unlocked`. `evp_pkey_reset_unlocked` will call
`memset` on `pk`, and then create a new lock and set `pk->lock` to point to
that new lock. I believe there are two problems with this.

The first is that after the call to `memset`, another thread would try to
acquire a lock for `NULL` as that is what the value of `pk->lock` would be
at that point.

The second issue is that after the new lock has been assigned to
`pk->lock`, that lock is different from the one currently locked so
another thread trying to acquire the lock will succeed which can lead to
strange behaviour. More details and a reproducer can be found in the
Refs link below.

This commit introduces a new function that is only called by
`evp_pkey_downgrade` which does not use `memset` but instead
"manually" sets the `EVP_PKEY` values to their default values, but does
not modify `pk->lock`. This could perhaps be updated to go back to only
having one function that is called for both `evp_pkey_downgrade` and
`EVP_PKEY_new` and only create a new lock if one does not already exist.

Refs:
https://github.com/danbev/learning-libcrypto/blob/master/notes/issues.md#openssl-investigationtroubleshooting  
https://github.com/nodejs/node/issues/29817

### CODE REVIEW

**Comment by mattcaswell** [comment]
I'm wondering why we have a lock at all? In line with most OpenSSL objects, an EVP_PKEY is not thread safe.

Ping @levitte.

**Comment by t8m** [comment]
Would it rather make sense to do the downgrade the other way around?
I.e. do the copy_downgraded to a temporary copy and then just copy back the contents if the downgrade succeeds (excluding the lock).

**Comment by mattcaswell** [comment]
Access to most of the fields in an EVP_PKEY is not controlled by the lock anyway. The lock seems to be primarily related to the reference counting.

**Comment by paulidale** [comment]
Reference counts are always associated with a lock.
This code plays with the locks, so it isn't a surprise.

File: crypto/evp/p_lib.c
Comment (levitte) [inline crypto/evp/p_lib.c:1439]: The only difference between this function and `evp_pkey_reset_unlocked` is that this one doesn't touch the lock, at all.  That's a bit over the top, I would rather see that this become a refactor of `evp_pkey_reset_unlocked`, and that the lock creation is moved back to `EVP_PKEY_new()`.

**Review by levitte — CHANGES_REQUESTED** [review]

**Comment by levitte** [comment]
I'm having some apprehensions about the new reset code...  addressing each field individually is a maintenance issue waiting to happen, as we will be prone forget to update the new reset function any time a new field is added in `EVP_PKEY`.  That is the single reason for the memset() method.

I do understand why a memset() isn't a best idea when a lock pointer is involved...  I wonder if a re-organization of EVP_PKEY could be a good idea, for example an internal structure to separate payload stuff from lock and count managerial stuff.
NOTE: *This is food for thought*, not something that I expect to happen in this PR!

File: crypto/evp/p_lib.c
Comment (danbev) [reply crypto/evp/p_lib.c:1439]: Sounds good, I'll make these changes. Thanks!

**Comment by paulidale** [comment]
What about copying the `EVP_PKEY` to a temporary.  Using `memset(2)` to clear everything, setting the lock in the copy and them copying back to the original?  It requires the lock pointer to be moved atomically which might be a bad assumption.

Something like this should work:
``` C
struct {
    ...
    CRYPTO_RWLOCK *lock;
    ...
} *p;

const size_t offset = (unsigned char *)&p->lock - (unsigned char *)p;    /* strictly it's a ptrdiff_t not size_t */

memset(p, 0, offset);
memset(&p->lock + sizeof(p->lock), 0, sizeof(*p) - offset - sizeof(p->lock));
```

**Comment by danbev** [comment]
@paulidale I'll take a closer look at your suggestion, but it might not be until tomorrow. Thanks

**Comment by paulidale** [comment]
It's only a suggestion, but I've reservations like @levitte.

**Comment by levitte** [comment]
> What about copying the `EVP_PKEY` to a temporary.  Using `memset(2)` to clear everything, setting the lock in the copy and them copying back to the original?

Doesn't that *exactly* reproduce the issue that this PR is trying to solve?

**Comment by paulidale** [comment]
> Doesn't that _exactly_ reproduce the issue that this PR is trying to solve?

The problem is that the original PKEY structure is being zeroed including the lock and then the lock is restored.  What I suggested doesn't do this.  That doesn't make it any less of a hack.

Selective `memset(2)` is better -- I think it will need to preserve the reference count and lock though.

**Comment by levitte** [comment]
Oh, now I see...

**Comment by danbev** [comment]
@paulidale I've added a commit with the changes you suggested (hope I did not misunderstand). One thing though was the second `memset` call which I had to change to be 1 instead of `sizeof(pk->lock)`. My reasoning for this can be found below.
<details>
<summary>memset reasoning</summary>

After the first call to memset, which will set all the values up to pk->lock to zero:                                                                        
```console                                                                      
(lldb) expr *pk                                                                 
(EVP_PKEY) $7 = {                                                               
  type = 0                                                                      
  save_type = 0                                                                 
  ameth = 0x0000000000000000                                                    
  engine = 0x0000000000000000                                                   
  pmeth_engine = 0x0000000000000000                                             
  pkey = {                                                                      
    ptr = 0x0000000000000000                                                    
    rsa = 0x0000000000000000                                                    
    dsa = 0x0000000000000000                                                    
    dh = 0x0000000000000000                                                     
    ec = 0x0000000000000000                                                     
    ecx = 0x0000000000000000                                                    
  }                                                                             
  references = 0                                                                
  lock = 0x00000000004520b0                                                     
  attributes = 0x0000000000000000                                               
  save_parameters = 1                                                           
  ex_data = {                                                                   
    ctx = 0x0000000000000000                                                    
    sk = 0x0000000000000000                                                     
  }                                                                             
  keymgmt = 0x000000000044fbb0                                                  
  keydata = 0x000000000045aea0                                                  
```                                                                             
And just take note of the address of pk->lock:                                  
```console                                                                      
(lldb) expr pk->lock                                                            
(CRYPTO_RWLOCK *) $3 = 0x00000000004520b0                                       
(lldb) memory read -c 1 -f x -s 8 &pk->lock                                     
0x00450a90: 0x00000000004520b0                                                  
```
Next we are going to use `memset` once again with the address of the lock       
plus the size of the pointer (which is 8)                                       
```console                                                                      
(lldb) expr &pk->lock + sizeof(pk->lock)                                        
(CRYPTO_RWLOCK **) $8 = 0x0000000000450ad0  ------------+                       
(lldb) expr sizeof(pk->lock)                            |                       
(unsigned long) $9 = 8                                  |                       
(lldb) memory read -c 20 -s 8 -f x pk                   |                       
0x00450a60: 0x0000000000000000 0x0000000000000000       |                       
0x00450a70: 0x0000000000000000 0x0000000000000000       |                       
0x00450a80: 0x0000000000000000 0x0000000000000000       |                       
0x00450a90: 0x00000000004520b0 0x0000000000000000       |                       
0x00450aa0: 0x0000000000000001 0x0000000000000000       |                       
0x00450ab0: 0x0000000000000000 0x000000000044fbb0       |                       
0x00450ac0: 0x000000000045aea0 0x0000000000000000       |                       
    +---------------------------------------------------+                       
    ↓                                                                           
0x00450ad0: 0x0000000000000000 0x0000000000000000                               
0x00450ae0: 0x0000000000000000 0x0000000000000000                               
0x00450af0: 0x0000000000000000 0x0000000000000000                               
```                                                                             
Notice that this skipped a bit further than expected. In the code we are adding 
8 to the address of pk->lock, but pk-lock is a pointer and perhaps it should    
only be adding 1?:                                                              
```console                                                                      
(lldb) memory read -c 1 -s 8 -f x '&pk->lock + 1'                               
0x00450a98: 0x0000000000000000                                                  
  ↑                                                                             
  +-------------------------------------------------+                           
0x00450a60: 0x0000000000000000 0x0000000000000000   |                           
0x00450a70: 0x0000000000000000 0x0000000000000000   |                           
0x00450a80: 0x0000000000000000 0x0000000000000000   |                           
                                      +-------------+                           
                                      ↓                                         
0x00450a90: 0x00000000004520b0 0x0000000000000000                               
0x00450aa0: 0x0000000000000001 0x0000000000000000                               
0x00450ab0: 0x0000000000000000 0x000000000044fbb0                               
0x00450ac0: 0x000000000045aea0 0x0000000000000000                               
0x00450ad0: 0x0000000000000000 0x0000000000000000                               
0x00450ae0: 0x0000000000000000 0x0000000000000000                               
0x00450af0: 0x0000000000000000 0x0000000000000000                               
```

</details>

File: crypto/evp/p_lib.c
Comment (t8m) [inline crypto/evp/p_lib.c:1392]: This is correct in terms of address computation. I am not sure it is strictly correct in regards to C standard. Isn't effectively dereferencing array (of a single element) past its bounds an undefined operation? But is memset() actually counted as dereference?

Not sure if `(unsigned char *)&pk->lock + sizeof(pk->lock) would make it any different though.
What about `memset((unsigned char *)pk + offset + sizeof(pk->lock), 0, sizeof(*pk) - offset - sizeof(pk->lock));` ?

**Review by t8m — CHANGES_REQUESTED** [review]

File: crypto/evp/p_lib.c
Comment (danbev) [reply crypto/evp/p_lib.c]: I've updated this now, thanks!

File: crypto/evp/p_lib.c
Comment (danbev) [reply crypto/evp/p_lib.c:1392]: Added it now, thanks

File: crypto/evp/p_lib.c
Comment (danbev) [reply crypto/evp/p_lib.c]: That worked like a charm, thanks again.

Comment (paulidale) [reply crypto/evp/p_lib.c]: I have a memory of the C standard allowing arrays to be indexed one position past their end without it becoming an undefined operation.

The alternative is better.

File: crypto/evp/p_lib.c
Comment (t8m) [inline crypto/evp/p_lib.c:1397]: Nit: Please amend the comment to something like: `EVP_PKEY_new uses zalloc so no need to call memset if pk->lock is NULL.`

Comment (danbev) [reply crypto/evp/p_lib.c]: Update now, thanks

**Review by t8m — DISMISSED** [review]
LGTM

File: crypto/evp/p_lib.c
Comment (levitte) [inline crypto/evp/p_lib.c:1420]: `ERR_raise(ERR_LIB_EVP, ERR_R_MALLOC_FAILURE);`

Please

File: crypto/evp/p_lib.c
Comment (levitte) [inline crypto/evp/p_lib.c:1986]: `/* |pk| now owns THE lock */` isn't fitting any more, is it?

File: crypto/evp/p_lib.c
Comment (kaduk) [reply crypto/evp/p_lib.c]: It is defined behavior to compute the value of a pointer that points one (but not two!) past the end of an object/array, yes.
Most of the text about (un)defined behavior with respect to this sort of access is written in terms of "object"s, and in this case one would say that the `evp_pkey_st` is the object, so I'm not really convinved there would be a problem in the old formulation either.

Another thing to note, though, is that the compiler is not exactly obligated to lay things out in memory in the order that the fields are declared in the structure definition.  In practice it mostly is, due to the requirement to be able to link with other objects, the strictures of platform ABI specs, etc., though there are some "hardening" compiler options that will randomize struct layouts.  It looks like we're currently only preserving the lock, not the refcount, but if we were trying to skip the memset for both fields we'd want to add some sort of check/assertion that the fields are laid out in the order we expect.

File: crypto/evp/p_lib.c
Comment (danbev) [reply crypto/evp/p_lib.c]: Thanks, fixed this now.

File: crypto/evp/p_lib.c
Comment (danbev) [reply crypto/evp/p_lib.c]: I've removed this now, thanks

**Comment by danbev** [comment]
The Travis CI build [failed](https://travis-ci.com/github/openssl/openssl/jobs/440537191) with the following error:
```console
An error occurred while generating the build script.
```
Would someone with permissions be able to re-run that job?

**Comment by levitte** [comment]
> Would someone with permissions be able to re-run that job?

It looks like someone did.

However, you will have to wait until #13450 is merged, and rebase then.  It fixes the other Travis failures we currently see.

**Comment by danbev** [comment]
> However, you will have to wait until #13450 is merged, and rebase then. It fixes the other Travis failures we currently see.

Thanks for the info and I'll rebase when that PRs has been merged.

**Comment by danbev** [comment]
@levitte Just a friendly ping that I've rebased this PR now that #13450 has landed.

**Review by levitte — APPROVED** [review]

**Review by t8m — APPROVED** [review]
Still approved.

**Comment by t8m** [comment]
Merged to master with amendment of the commit message to reflect the real change after all the squashed commits.

**Comment by t8m** [comment]
Thank you for the contribution.

**Pull request closed without merging**
